io/iommu/iommu_tests.py

```diff
@@ -0,0 +1,373 @@
+"""
+
+import os
+from avocado import Test
+from avocado.utils import process
+from avocado.utils import pci, genio
+from avocado.utils.software_manager.manager import SoftwareManager
+from avocado.utils import wait
+
+
+def change_domain_check(dom, pci_addr, def_dom):
+    """
+    Check if the domain changed successfully to "dom" for "pci_addr"
+
+    :param dom: domain type
+    :param def_dom: default domain of pci device(pci_addr)
+    :param pci_addr: pci device
+    return: bool
+    """
+    output = genio.read_one_line(
+            f"/sys/bus/pci/devices/{pci_addr}/iommu_group/type")
+    out = output.rsplit(None, 1)[-1]
+    if (dom not in ("auto", out)) or (dom == "auto" and out != def_dom):
+        return False
+    return True
+
+
+def reset_check(pci_addr):
+    """
+    Check if reset for "pci_addr" is successful
+
+    :param pci_addr: pci device
+    return: bool
+    """
+    cmd = f"lspci -vvs {pci_addr}"
+    output = process.run(cmd, ignore_status=True, shell=True).stdout_text
+    if output != "":
+        return False
+    return True
+
+
+def rescan_check(pci_addr):
+    """
+    Check if rescan for pci_addr is successful
+
+    :param pci_addr: pci device
+    return: bool
+    """
+    cmd = f"lspci -vvs {pci_addr}"
+    output = process.run(cmd, ignore_status=True, shell=True).stdout_text
+    if output == "":
+        return False
+    return True
+
+
+class IommuTest(Test):
+
+    """
+    Various combination of unbind, bind, change iommu group domain type,
+    reset and rescan is used to form sub-tests that test and exercise iommu
+    code.
+
+    :param device: Name of the pci device
+    """
+
+    def setUp(self):
+        """
+        Setup the device
+        """
+        self.pci_devices = self.params.get('pci_devices', default=None)
+        self.count = int(self.params.get('count', default=1))
+        self.domains = ["DMA", "DMA-FQ", "identity", "auto"]
+        if not self.pci_devices:
+            self.cancel("No pci device given")
+        smm = SoftwareManager()
+        if not smm.check_installed("pciutils") and not smm.install("pciutils"):
+            self.cancel("pciutils package not found and installing failed")
+        # Check the number of devices in iommu-group for pci device passed.
+        for pci_addr in self.pci_devices.split(" "):
+            driver = pci.get_driver(pci_addr)
+            if driver is None:
+                self.cancel("Device passed is not bound to any driver")
+            lst = os.listdir(f'/sys/bus/pci/drivers/{driver}/{pci_addr}/'
+                             'iommu_group/devices/')
+            if len(lst) != 1:
+                self.cancel(f"{pci_addr} belongs to iommu group having more "
+                            "than one device but system doesnot support "
+                            "domain type change for such device")
+        cmd = "dmesg -C"
+        process.run(cmd, ignore_status=True, shell=True, sudo=True)
```

Comment:
please dmesg library from avocado https://github.com/avocado-framework/avocado/blob/master/avocado/utils/dmesg.py

Reply:
>

Avocado function to clear dmesg uses "dmesg -c" command which will first show the dmesg then clears it. So for two reason not included that
1) This will show all the system dmesgs that are there before the start of first subtest case.
2) Do we require all the dmesgs for each subtest to be printed. As of now we are mainly concerned with dmesgs with levels=warning, critical, error and alert. These if present are printed and test fails.

So we used instead "dmesg -C" which simply clears the dmesg. Thats the reason of not using inbuilt function in avocado.

Reply:
sure in that case we should be enhancing the library may have function argument  in the library which can decide this scenario , please let me know you need more info on this as i love to see this should be implement in library

Reply:
Sure.